File: Task1B/product.cpp
#include "product.h"
#include "supplier.h"

// Default constructor
Product::Product() {
    this->name = "";
    this->price = 0;
    this->supplier = NULL;
    this->volume = 0;
}

// Parameterised constructor
Product::Product(string name, int price, Supplier* supplier, int volume, int productID) {
    this->name = name;
    this->price = price;
    this->supplier = supplier;
    this->volume = volume;
    this->productID = productID;
}

// Display product information
void Product::display_product_info() {
    cout << "Name: " << this->name << endl;
    cout << "Price: " << this->price << endl;
    cout << "Supplier: " << endl;
    this->supplier->viewSupplierInfo();
    cout << "Volume: " << this->volume << endl;
    cout << "Product ID: " << this->productID << endl;
}

// Getters and setters
string Product::getName() {
    return this->name;
}

int Product::getPrice() {
    return this->price;
}

Supplier* Product::getSupplier() {
    return this->supplier;
}

int Product::getVolume() {
    return this->volume;
}

int Product::getProductID() {
    return this->productID;
}

void Product::setName(string name) {
    this->name = name;
}

void Product::setPrice(int price) {
    this->price = price;
}

void Product::setSupplier(Supplier* supplier) {
    this->supplier = supplier;
}

void Product::setVolume(int volume) {
    this->volume = volume;
}


void Product::setProductID(int productID) {
    this->productID = productID;
}

// Save product to file
void Product::save(ostream& out) const {
    out << name << "," << price << "," << volume << "," << productID << ",";
    supplier->save(out);
}

// Load product from file
void Product::load(istream& in) {
    string line;
    getline(in, line);
    stringstream ss(line);

    string token;
    getline(ss, token, ',');
    name = token;

    getline(ss, token, ',');
    price = stoi(token);

    getline(ss, token, ',');
    volume = stoi(token);

    getline(ss, token, ',');
    productID = stoi(token);

    getline(ss, token, ',');
    string supplierName = token;

    getline(ss, token, ',');
    int supplierID = stoi(token);
    supplier = new Supplier(supplierName, supplierID);
}

// Destructor
Product::~Product() {
    delete supplier;
}

File: Task4A_B/task4a_b/src/main/java/com/cps2004/Product.java
package com.cps2004;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.node.ObjectNode;

public abstract class Product {
    private String name;
    private int id;
    private Supplier supplier;
    private int price;
    private int volume; 

    // Default constructor
    public Product() {
        name = "";
        id = 0;
        supplier = new Supplier();
        price = 0;
        volume = 0;
    }

    // Parameterized constructor
    public Product(String name, int id, Supplier supplier, int price, int volume) {
        this.name = name;
        this.id = id;
        this.supplier = supplier;
        this.price = price;
        this.volume = volume;
    }

    // Getters and setters
    public String getName() { return name; }
    public int getId() { return id; }
    public Supplier getSupplier() { return supplier; }
    public int getPrice() { return price; }
    public int getVolume() { return volume; }
    public void setName(String name) { this.name = name; }
    public void setId(int id) { this.id = id; }
    public void setSupplier(Supplier supplier) { this.supplier = supplier; }
    public void setPrice(int price) { this.price = price; }
    public void setVolume(int volume) { this.volume = volume; }

    // Calculate discounted price
    public abstract double calculate_discount(String month, int quantity);

    // Display product information
    public void display_product_info() {
        System.out.println("Product name: " + name);
        System.out.println("Product ID: " + id);
        System.out.println("Supplier name: " + supplier.getName());
        System.out.println("Supplier ID: " + supplier.getId());
        System.out.println("Product price: " + price);
        System.out.println("Product volume: " + volume);
    }
}
